feat(router): Cascading route support (#16416)
angular/angular: aio/content/guide/router.md

A guard's return value controls the router's behavior:

* If it returns `true`, the navigation process continues.
* If it returns `false`, the navigation process stops and the user stays put
**unless** the `cascade` option is enabled on [`ExtraOptions`](api/router/ExtraOptions). 

> When `cascade` is enabled the router will continue searching for additional viable routes.

Reviewer: ```suggestion
When `cascade` is enabled the router continues searching for additional viable routes.
```

* If it returns a `UrlTree`, the current navigation cancels and a new navigation is initiated to the `UrlTree` returned.

<div class="alert is-helpful">

**Note:** The guard can also tell the router to navigate elsewhere, effectively canceling the current navigation. When

> doing so inside a guard, the guard should return `false`; However, when the `cascade` option under 

Reviewer: Is it still correct without "should"? (Same thought for the next "should".)

Author: That's the original wording in the docs I believe. i.e. saying -- hey, if you call router.navigate() inside your guard then you "should return 'false'" which is really an efficiency to bypass the in-process navigation. That wording is not related to this code...my change starts at "However, ...."

Author: as I look at this I believe "should" is necessary here...it is a opinionated recommendation to the guard implementor.

Reviewer: Thanks for the explanation. :)

The router checks the `CanDeactivate` and `CanActivateChild` guards first, from the deepest child route to the top.
Then it checks the `CanActivate` guards from the top down to the deepest child route. If the feature module
is loaded asynchronously, the `CanLoad` guard is checked before the module is loaded.

> If _any_ guard returns false, pending guards that have not completed will also be canceled.

Reviewer: ```suggestion
If _any_ guard returns false, pending guards that have not completed are also canceled.
```

Author: Committed. This is unrelated to this PR however as that line I believe was pre-existing.

Reviewer: Ah, gotcha. I should have turned off whitespace! (I've been trying to remember that 🙃)

> Additionally, the entire navigation sequence itself will be canceled **unless** the `cascade` option in

Reviewer: ```suggestion
Additionally, the entire navigation sequence itself is canceled **unless** the `cascade` option in
```
